In [2]:
import pandas as pd
import numpy as np
import zipfile
from scipy import stats

In [3]:
df_desc = pd.read_csv('content/data_dictionary.csv')
df_desc

Instrument  \
0                           Identifier   
1                         Demographics   
2                         Demographics   
3                         Demographics   
4   Children's Global Assessment Scale   
..                                 ...   
76             Sleep Disturbance Scale   
77             Sleep Disturbance Scale   
78             Sleep Disturbance Scale   
79                        Internet Use   
80                        Internet Use   

                                     Field                       Description  \
0                                       id                  Participant's ID   
1                Basic_Demos-Enroll_Season              Season of enrollment   
2                          Basic_Demos-Age                Age of participant   
3                          Basic_Demos-Sex                Sex of participant   
4                              CGAS-Season           Season of participation   
..                                     ...                               ...   
76                              SDS-Season           Season of participation   
77                       SDS-SDS_Total_Raw                   Total Raw Score   
78                         SDS-SDS_Total_T                     Total T-Score   
79                     PreInt_EduHx-Season           Season of participation   
80  PreInt_EduHx-computerinternet_hoursday  Hours of using computer/internet   

               Type                        Values  \
0               str                           NaN   
1               str  Spring, Summer, Fall, Winter   
2             float                           NaN   
3   categorical int                           0,1   
4               str  Spring, Summer, Fall, Winter   
..              ...                           ...   
76              str  Spring, Summer, Fall, Winter   
77              int                           NaN   
78              int                           NaN   
79              str  Spring, Summer, Fall, Winter   
80  categorical int                       0,1,2,3   

                                         Value Labels  
0                                                 NaN  
1                                                 NaN  
2                                                 NaN  
3                                    0=Male, 1=Female  
4                                                 NaN  
..                                                ...  
76                                                NaN  
77                                                NaN  
78                                                NaN  
79                                                NaN  
80  0=Less than 1h/day, 1=Around 1h/day, 2=Around ...  

[81 rows x 6 columns]

In [4]:
df = pd.read_csv('content/train.csv')
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1            NaN              NaN            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956         NaN              NaN          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959         NaN              NaN          Winter           NaN   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0                46.0             50.8  ...             4.0             2.0   
1                48.0             46.0  ...             0.0             0.0   
2                56.5             75.6  ...             2.0             1.0   
3                56.0             81.6  ...             3.0             4.0   
4                 NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3955             59.5             82.4  ...             1.0             1.0   
3956             53.5             76.4  ...             NaN             NaN   
3957             60.0            109.8  ...             1.0             0.0   
3958             70.7             87.0  ...             1.0             1.0   
3959              NaN              NaN  ...             NaN             NaN   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season  SDS-SDS_Total_Raw  \
0                4.0               55.0        NaN                NaN   
1                0.0                0.0       Fall               46.0   
2                1.0               28.0       Fall               38.0   
3                1.0               44.0     Summer               31.0   
4                NaN                NaN        NaN                NaN   
...              ...                ...        ...                ...   
3955             0.0               32.0     Winter               35.0   
3956             NaN                NaN        NaN                NaN   
3957             1.0               31.0     Winter               56.0   
3958             1.0               19.0     Spring               33.0   
3959             NaN                NaN        NaN                NaN   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                 NaN                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                 NaN                  NaN   
...               ...                  ...   
3955             50.0                 Fall   
3956              NaN               Winter   
3957             77.0         

In [5]:
df.shape

(3960, 82)

# Before we begin...

Our data has categorical information, true numerical information, and categorical information stored as numbers

In [6]:
df_desc.head() # look at type column

Instrument                      Field  \
0                          Identifier                         id   
1                        Demographics  Basic_Demos-Enroll_Season   
2                        Demographics            Basic_Demos-Age   
3                        Demographics            Basic_Demos-Sex   
4  Children's Global Assessment Scale                CGAS-Season   

               Description             Type                        Values  \
0         Participant's ID              str                           NaN   
1     Season of enrollment              str  Spring, Summer, Fall, Winter   
2       Age of participant            float                           NaN   
3       Sex of participant  categorical int                           0,1   
4  Season of participation              str  Spring, Summer, Fall, Winter   

       Value Labels  
0               NaN  
1               NaN  
2               NaN  
3  0=Male, 1=Female  
4               NaN

In [7]:
df_desc['Type'].unique()

array(['str', 'float', 'categorical int', 'int'], dtype=object)

In [8]:
df_desc[df_desc['Field'] == 'sii'] # worthy to note that the description table doesn't have a row for the output sii

Empty DataFrame
Columns: [Instrument, Field, Description, Type, Values, Value Labels]
Index: []

In [9]:
# categoricals = df.dtypes[df.dtypes == 'object'].index
categoricals = pd.Index(df_desc[df_desc['Type'] == 'str']['Field'])
categoricals # seasonal data + id

Index(['id', 'Basic_Demos-Enroll_Season', 'CGAS-Season', 'Physical-Season',
       'Fitness_Endurance-Season', 'FGC-Season', 'BIA-Season', 'PAQ_A-Season',
       'PAQ_C-Season', 'PCIAT-Season', 'SDS-Season', 'PreInt_EduHx-Season'],
      dtype='object', name='Field')

In [10]:
# allnumericals = df.dtypes[df.dtypes != 'object'].index
allnumericals = pd.Index(df_desc[df_desc['Type'] != 'str']['Field'])
allnumericals

Index(['Basic_Demos-Age', 'Basic_Demos-Sex', 'CGAS-CGAS_Score', 'Physical-BMI',
       'Physical-Height', 'Physical-Weight', 'Physical-Waist_Circumference',
       'Physical-Diastolic_BP', 'Physical-HeartRate', 'Physical-Systolic_BP',
       'Fitness_Endurance-Max_Stage', 'Fitness_Endurance-Time_Mins',
       'Fitness_Endurance-Time_Sec', 'FGC-FGC_CU', 'FGC-FGC_CU_Zone',
       'FGC-FGC_GSND', 'FGC-FGC_GSND_Zone', 'FGC-FGC_GSD', 'FGC-FGC_GSD_Zone',
       'FGC-FGC_PU', 'FGC-FGC_PU_Zone', 'FGC-FGC_SRL', 'FGC-FGC_SRL_Zone',
       'FGC-FGC_SRR', 'FGC-FGC_SRR_Zone', 'FGC-FGC_TL', 'FGC-FGC_TL_Zone',
       'BIA-BIA_Activity_Level_num', 'BIA-BIA_BMC', 'BIA-BIA_BMI',
       'BIA-BIA_BMR', 'BIA-BIA_DEE', 'BIA-BIA_ECW', 'BIA-BIA_FFM',
       'BIA-BIA_FFMI', 'BIA-BIA_FMI', 'BIA-BIA_Fat', 'BIA-BIA_Frame_num',
       'BIA-BIA_ICW', 'BIA-BIA_LDM', 'BIA-BIA_LST', 'BIA-BIA_SMM',
       'BIA-BIA_TBW', 'PAQ_A-PAQ_A_Total', 'PAQ_C-PAQ_C_Total',
       'PCIAT-PCIAT_01', 'PCIAT-PCIAT_02', 'PCIAT-PCIAT_03

In [11]:
# numcategoricals = allnumericals[(df[allnumericals].min() == 0) & (df[allnumericals].max() <= 5)]
numcategoricals = pd.Index(df_desc[df_desc['Type'] == 'categorical int']['Field']).append(pd.Index(['sii'],name='Field'))
numcategoricals

Index(['Basic_Demos-Sex', 'FGC-FGC_CU_Zone', 'FGC-FGC_GSND_Zone',
       'FGC-FGC_GSD_Zone', 'FGC-FGC_PU_Zone', 'FGC-FGC_SRL_Zone',
       'FGC-FGC_SRR_Zone', 'FGC-FGC_TL_Zone', 'BIA-BIA_Activity_Level_num',
       'BIA-BIA_Frame_num', 'PCIAT-PCIAT_01', 'PCIAT-PCIAT_02',
       'PCIAT-PCIAT_03', 'PCIAT-PCIAT_04', 'PCIAT-PCIAT_05', 'PCIAT-PCIAT_06',
       'PCIAT-PCIAT_07', 'PCIAT-PCIAT_08', 'PCIAT-PCIAT_09', 'PCIAT-PCIAT_10',
       'PCIAT-PCIAT_11', 'PCIAT-PCIAT_12', 'PCIAT-PCIAT_13', 'PCIAT-PCIAT_14',
       'PCIAT-PCIAT_15', 'PCIAT-PCIAT_16', 'PCIAT-PCIAT_17', 'PCIAT-PCIAT_18',
       'PCIAT-PCIAT_19', 'PCIAT-PCIAT_20',
       'PreInt_EduHx-computerinternet_hoursday', 'sii'],
      dtype='object', name='Field')

In [12]:
# numericals = allnumericals[~((df[allnumericals].min() == 0) & (df[allnumericals].max() <= 5))]
numericals = pd.Index(df_desc[(df_desc['Type'] == 'float') | (df_desc['Type'] == 'int')]['Field'])
numericals

Index(['Basic_Demos-Age', 'CGAS-CGAS_Score', 'Physical-BMI', 'Physical-Height',
       'Physical-Weight', 'Physical-Waist_Circumference',
       'Physical-Diastolic_BP', 'Physical-HeartRate', 'Physical-Systolic_BP',
       'Fitness_Endurance-Max_Stage', 'Fitness_Endurance-Time_Mins',
       'Fitness_Endurance-Time_Sec', 'FGC-FGC_CU', 'FGC-FGC_GSND',
       'FGC-FGC_GSD', 'FGC-FGC_PU', 'FGC-FGC_SRL', 'FGC-FGC_SRR', 'FGC-FGC_TL',
       'BIA-BIA_BMC', 'BIA-BIA_BMI', 'BIA-BIA_BMR', 'BIA-BIA_DEE',
       'BIA-BIA_ECW', 'BIA-BIA_FFM', 'BIA-BIA_FFMI', 'BIA-BIA_FMI',
       'BIA-BIA_Fat', 'BIA-BIA_ICW', 'BIA-BIA_LDM', 'BIA-BIA_LST',
       'BIA-BIA_SMM', 'BIA-BIA_TBW', 'PAQ_A-PAQ_A_Total', 'PAQ_C-PAQ_C_Total',
       'PCIAT-PCIAT_Total', 'SDS-SDS_Total_Raw', 'SDS-SDS_Total_T'],
      dtype='object', name='Field')

In [13]:
len(categoricals) + len(numcategoricals) + len(numericals), df.shape[1]

(82, 82)

# Cleaning up the Zeros

Some of the data is filled in with 0 when it really should be `NaN`

In [14]:
df.describe().loc['min'][df.describe().loc['min'] < 0].index

Index(['BIA-BIA_BMC', 'BIA-BIA_FMI', 'BIA-BIA_Fat'], dtype='object')

These columns have a negative minimum. Is this possible?

In [15]:
df_desc[df_desc['Field'].isin(df.describe().loc['min'][df.describe().loc['min'] < 0].index)]

Instrument        Field           Description   Type  \
35  Bio-electric Impedance Analysis  BIA-BIA_BMC  Bone Mineral Content  float   
42  Bio-electric Impedance Analysis  BIA-BIA_FMI        Fat Mass Index  float   
43  Bio-electric Impedance Analysis  BIA-BIA_Fat   Body Fat Percentage  float   

   Values Value Labels  
35    NaN          NaN  
42    NaN          NaN  
43    NaN          NaN

You cannot have negative bone mineral content, negative fat mass index, or negative body fat percentage. We should replace the negative values with `NaN`.

In [16]:
replaceNegatives = df.describe().loc['min'][df.describe().loc['min'] < 0].index
replaceNegatives

Index(['BIA-BIA_BMC', 'BIA-BIA_FMI', 'BIA-BIA_Fat'], dtype='object')

In [17]:
categoricals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index)

Index([], dtype='object')

In [18]:
numericals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index)

Index(['Physical-BMI', 'Physical-Weight', 'Physical-Diastolic_BP',
       'Physical-Systolic_BP', 'Fitness_Endurance-Max_Stage',
       'Fitness_Endurance-Time_Mins', 'Fitness_Endurance-Time_Sec',
       'FGC-FGC_CU', 'FGC-FGC_GSND', 'FGC-FGC_GSD', 'FGC-FGC_PU',
       'FGC-FGC_SRL', 'FGC-FGC_SRR', 'FGC-FGC_TL', 'PCIAT-PCIAT_Total'],
      dtype='object')

These numerical columns have a minimum of 0. Is this possible?  
For most of them, they should not have values of 0. For example `Physical-Weight` of 0 does not mean anything, so we should replace 0s with `NaN`.  
However, the one exception is the column `PCIAT-PCIAT_Total`, which is the total score to the Parent-Child Internet Addiction Test, where it is possible to have a score of 0.

In [19]:
numcategoricals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index)

Index(['Basic_Demos-Sex', 'FGC-FGC_CU_Zone', 'FGC-FGC_PU_Zone',
       'FGC-FGC_SRL_Zone', 'FGC-FGC_SRR_Zone', 'FGC-FGC_TL_Zone',
       'PCIAT-PCIAT_01', 'PCIAT-PCIAT_02', 'PCIAT-PCIAT_03', 'PCIAT-PCIAT_04',
       'PCIAT-PCIAT_05', 'PCIAT-PCIAT_06', 'PCIAT-PCIAT_07', 'PCIAT-PCIAT_08',
       'PCIAT-PCIAT_09', 'PCIAT-PCIAT_10', 'PCIAT-PCIAT_11', 'PCIAT-PCIAT_12',
       'PCIAT-PCIAT_13', 'PCIAT-PCIAT_14', 'PCIAT-PCIAT_15', 'PCIAT-PCIAT_16',
       'PCIAT-PCIAT_17', 'PCIAT-PCIAT_18', 'PCIAT-PCIAT_19', 'PCIAT-PCIAT_20',
       'PreInt_EduHx-computerinternet_hoursday', 'sii'],
      dtype='object')

These categorical columns have a minimum of 0. Is this possible?  
Most likely yes, since even though they are recorded as numbers, they are 'categorical'. For example, `Basic_Demos-Sex` is recorded as 0 and 1.

In [20]:
replaceZeros = numericals.intersection(df.describe().loc['min'][df.describe().loc['min'] == 0].index).drop('PCIAT-PCIAT_Total')
replaceZeros

Index(['Physical-BMI', 'Physical-Weight', 'Physical-Diastolic_BP',
       'Physical-Systolic_BP', 'Fitness_Endurance-Max_Stage',
       'Fitness_Endurance-Time_Mins', 'Fitness_Endurance-Time_Sec',
       'FGC-FGC_CU', 'FGC-FGC_GSND', 'FGC-FGC_GSD', 'FGC-FGC_PU',
       'FGC-FGC_SRL', 'FGC-FGC_SRR', 'FGC-FGC_TL'],
      dtype='object')

In [21]:
df_desc[df_desc['Field'].isin(replaceZeros)]

Instrument                        Field  \
7                  Physical Measures                 Physical-BMI   
9                  Physical Measures              Physical-Weight   
11                 Physical Measures        Physical-Diastolic_BP   
13                 Physical Measures         Physical-Systolic_BP   
15  FitnessGram Vitals and Treadmill  Fitness_Endurance-Max_Stage   
16  FitnessGram Vitals and Treadmill  Fitness_Endurance-Time_Mins   
17  FitnessGram Vitals and Treadmill   Fitness_Endurance-Time_Sec   
19                 FitnessGram Child                   FGC-FGC_CU   
21                 FitnessGram Child                 FGC-FGC_GSND   
23                 FitnessGram Child                  FGC-FGC_GSD   
25                 FitnessGram Child                   FGC-FGC_PU   
27                 FitnessGram Child                  FGC-FGC_SRL   
29                 FitnessGram Child                  FGC-FGC_SRR   
31                 FitnessGram Child                   FGC-FGC_TL   

                           Description   Type Values Value Labels  
7             Body Mass Index (kg/m^2)  float    NaN          NaN  
9                         Weight (lbs)  float    NaN          NaN  
11                 Diastolic BP (mmHg)    int    NaN          NaN  
13                  Systolic BP (mmHg)    int    NaN          NaN  
15               Maximum stage reached    int    NaN          NaN  
16       Exact time completed: Minutes    int    NaN          NaN  
17       Exact time completed: Seconds    int    NaN          NaN  
19                       Curl up total    int    NaN          NaN  
21  Grip Strength total (non-dominant)  float    NaN          NaN  
23      Grip Strength total (dominant)  float    NaN          NaN  
25                       Push-up total    int    NaN          NaN  
27       Sit & Reach total (left side)  float    NaN          NaN  
29      Sit & Reach total (right side)  float    NaN          NaN  
31                    Trunk lift total    int    NaN          NaN

A value of 0 is impossible for BMI, weight, blood pressure, or any fitness test that counts reps.

We should replace these 0 values with `NaN`.

In [22]:
df[replaceNegatives] = df[replaceNegatives].mask(df[replaceNegatives] <= 0)

In [23]:
(df[replaceNegatives] <= 0).sum().sum()

0

In [24]:
df[replaceZeros] = df[replaceZeros].mask(df[replaceZeros] <= 0)

In [25]:
(df[replaceZeros] <= 0).sum().sum()

0

In [26]:
df.isnull().sum().sum() / (df.shape[0] * df.shape[1])

0.4108770633160877

41% of the data is now null

In [27]:
# df[(df[df.dtypes.index[df.dtypes != 'object']] < 0).any(axis=1)]

In [28]:
# df[df.dtypes.index[df.dtypes != 'object']] < 0

In [29]:
# df.mask(df[df.dtypes.index[df.dtypes != 'object']] < 0, np.nan)
# df[df.dtypes.index[df.dtypes != 'object']].mask(df[df.dtypes.index[df.dtypes != 'object']] < 0, np.nan)

## Removing columns in which more than 80% of the rows are null

In [30]:
(df.isnull().sum() > len(df)*0.8).sum()

5

There are 5 columns who have >80% missing values.

In [31]:
list(df.columns[df.isnull().sum() > len(df)*0.8])

['Fitness_Endurance-Max_Stage',
 'Fitness_Endurance-Time_Mins',
 'Fitness_Endurance-Time_Sec',
 'PAQ_A-Season',
 'PAQ_A-PAQ_A_Total']

In [32]:
df_desc[df_desc['Field'].isin(df.columns[df.isnull().sum() > len(df)*0.8])]

Instrument  \
15               FitnessGram Vitals and Treadmill   
16               FitnessGram Vitals and Treadmill   
17               FitnessGram Vitals and Treadmill   
50  Physical Activity Questionnaire (Adolescents)   
51  Physical Activity Questionnaire (Adolescents)   

                          Field                           Description   Type  \
15  Fitness_Endurance-Max_Stage                 Maximum stage reached    int   
16  Fitness_Endurance-Time_Mins         Exact time completed: Minutes    int   
17   Fitness_Endurance-Time_Sec         Exact time completed: Seconds    int   
50                 PAQ_A-Season               Season of participation    str   
51            PAQ_A-PAQ_A_Total  Activity Summary Score (Adolescents)  float   

                          Values Value Labels  
15                           NaN          NaN  
16                           NaN          NaN  
17                           NaN          NaN  
50  Spring, Summer, Fall, Winter          NaN  
51                           NaN          NaN

In [33]:
"""
filter_null = df.isnull().sum().sort_values(ascending = True) <= (len(df) / 2)
filter_null = filter_null[filter_null == True]
list_of_columns_less_null = filter_null.index.to_list()
df_filter_null = df[list_of_columns_less_null]
df_filter_null.isna().sum().sum() / len(df_filter_null)
len(df_filter_null)
"""
df = df.drop(columns=df.columns[df.isnull().sum() > len(df)*0.8])
df.head()

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0  00008ff9                      Fall                5                0   
1  000fd460                    Summer                9                0   
2  00105258                    Summer               10                1   
3  00115b9f                    Winter                9                0   
4  0016bb22                    Spring               18                1   

  CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  Physical-Height  \
0      Winter             51.0            Fall     16.877316             46.0   
1         NaN              NaN            Fall     14.035590             48.0   
2        Fall             71.0            Fall     16.648696             56.5   
3        Fall             71.0          Summer     18.292347             56.0   
4      Summer              NaN             NaN           NaN              NaN   

   Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  PCIAT-PCIAT_20  \
0             50.8  ...             4.0             2.0             4.0   
1             46.0  ...             0.0             0.0             0.0   
2             75.6  ...             2.0             1.0             1.0   
3             81.6  ...             3.0             4.0             1.0   
4              NaN  ...             NaN             NaN             NaN   

   PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  SDS-SDS_Total_T  \
0               55.0        NaN               NaN              NaN   
1                0.0       Fall              46.0             64.0   
2               28.0       Fall              38.0             54.0   
3               44.0     Summer              31.0             45.0   
4                NaN        NaN               NaN              NaN   

   PreInt_EduHx-Season  PreInt_EduHx-computerinternet_hoursday  sii  
0                 Fall                                     3.0  2.0  
1               Summer                                     0.0  0.0  
2               Summer                                     2.0  0.0  
3               Winter                                     0.0  1.0  
4                  NaN                                     NaN  NaN  

[5 rows x 77 columns]

In [34]:
df.isnull().sum().sum() / (df.shape[0] * df.shape[1])

0.38279876688967596

38% of the data is now null.

In [35]:
(df.isnull().sum(axis=1) > df.shape[1]*0.8).sum()

756

756 rows of the data have more than 80% of the attributes missing! Consider removing those rows.  
On the other hand, we also have time series data as well for some of the rows, so we can wait and see later on if we really need to remove the rows that have a lot of missing values.

In [36]:
numericals = numericals.drop(['Fitness_Endurance-Max_Stage','Fitness_Endurance-Time_Mins','Fitness_Endurance-Time_Sec','PAQ_A-PAQ_A_Total'])

In [37]:
categoricals = categoricals.drop(['PAQ_A-Season'])

# Removing the outliers

In [38]:
df.describe()

Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  Physical-BMI  \
count      3960.000000      3960.000000      2421.000000   3015.000000   
mean         10.433586         0.372727        65.454771     19.376812   
std           3.574648         0.483591        22.341862      5.034191   
min           5.000000         0.000000        25.000000      8.522436   
25%           8.000000         0.000000        59.000000     15.890526   
50%          10.000000         0.000000        65.000000     17.950925   
75%          13.000000         1.000000        75.000000     21.588631   
max          22.000000         1.000000       999.000000     59.132048   

       Physical-Height  Physical-Weight  Physical-Waist_Circumference  \
count      3027.000000      3015.000000                    898.000000   
mean         55.946713        90.840060                     27.278508   
std           7.473764        43.161374                      5.567287   
min          33.000000        31.800000                     18.000000   
25%          50.000000        58.200000                     23.000000   
50%          55.000000        77.800000                     26.000000   
75%          62.000000       114.300000                     30.000000   
max          78.500000       315.000000                     50.000000   

       Physical-Diastolic_BP  Physical-HeartRate  Physical-Systolic_BP  ...  \
count            2953.000000         2967.000000           2953.000000  ...   
mean               69.672536           81.597236            117.022689  ...   
std                13.553022           13.665196             16.927687  ...   
min                11.000000           27.000000             49.000000  ...   
25%                61.000000           72.000000            107.000000  ...   
50%                68.000000           81.000000            114.000000  ...   
75%                76.000000           90.500000            125.000000  ...   
max               179.000000          138.000000            203.000000  ...   

       PCIAT-PCIAT_16  PCIAT-PCIAT_17  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
count     2728.000000     2725.000000     2728.000000     2730.000000   
mean         1.452346        1.627890        1.613636        1.158974   
std          1.495600        1.445622        1.529178        1.343661   
min          0.000000        0.000000        0.000000        0.000000   
25%          0.000000        0.000000        0.000000        0.000000   
50%          1.000000        1.000000        1.000000        1.000000   
75%          2.000000        3.000000        2.000000        2.000000   
max          5.000000        5.000000        5.000000        5.000000   

       PCIAT-PCIAT_20  PCIAT-PCIAT_Total  SDS-SDS_Total_Raw  SDS-SDS_Total_T  \
count     2733.000000        2736.000000        2609.000000      2606.000000   
mean         0.943652          27.896199          41.088923        57.763622   
std          1.185460          20.338853          10.427433        13.196091   
min          0.000000           0.000000          17.000000        38.000000   
25%          0.000000          12.000000          33.000000        47.000000   
50%          1.000000          26.000000          39.000000        55.000000   
75%          1.000000          41.000000          46.000000        64.000000   
max          5.000000          93.000000          96.000000       100.000000   

       PreInt_EduHx-computerinternet_hoursday          sii  
count                             3301.000000  2736.000000  
mean                                 1.060588     0.580409  
std                                  1.094875     0.771122  
min                                  0.000000     0.000000  
25%                                  0.000000     0.000000  
50%                                  1.000000     0.000000  
75%                                  2.000000     1.000000  
max                                  3.000000     3.000000  

[8 rows x 66 columns]

In [39]:
df[(np.abs(stats.zscore(df[numericals], nan_policy='omit')) >= 3).any(axis=1)]

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
6     0038ba98                      Fall               10                0   
51    02cebf33                    Spring               12                1   
59    0380e709                    Summer               17                0   
87    04df895a                    Winter               17                0   
193   0acac70c                    Spring               18                0   
...        ...                       ...              ...              ...   
3876  f9ca9ff6                    Winter               11                1   
3903  fb9b060f                    Summer               16                1   
3922  fcd9a6bb                    Spring                6                0   
3946  feb53635                    Winter               14                0   
3952  ff60112d                    Summer               15                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
6            NaN              NaN            Fall     19.660760   
51        Spring             77.0          Summer     19.751301   
59           NaN              NaN          Summer     24.446886   
87           NaN              NaN          Winter     26.630764   
193         Fall             65.0          Summer     21.485616   
...          ...              ...             ...           ...   
3876         NaN              NaN          Spring     16.319643   
3903        Fall             65.0          Summer     21.266964   
3922      Summer             69.0          Spring     16.195145   
3946      Summer             70.0          Spring     19.036779   
3952      Spring             40.0          Winter     26.364710   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
6               55.00             84.6  ...             4.0             1.0   
51              62.00            108.0  ...             3.0             3.0   
59              68.00            160.8  ...             4.0             2.0   
87              77.00            224.6  ...             NaN             NaN   
193             70.15            150.4  ...             0.0             3.0   
...               ...              ...  ...             ...             ...   
3876            56.00             72.8  ...             NaN             NaN   
3903            67.00            135.8  ...             3.0             2.0   
3922            44.00             44.6  ...             1.0             0.0   
3946            69.50            130.8  ...             4.0             4.0   
3952            70.50            186.4  ...             1.0             1.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
6                0.0               20.0     Winter              27.0   
51               3.0               46.0     Summer              82.0   
59               2.0               56.0        NaN               NaN   
87               NaN                NaN        NaN               NaN   
193              4.0               34.0     Summer              33.0   
...              ...                ...        ...               ...   
3876             NaN                NaN        NaN               NaN   
3903             1.0               52.0     Summer              68.0   
3922             0.0               14.0     Spring              38.0   
3946             4.0               77.0     Spring              51.0   
3952             1.0               35.0     Spring              66.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
6                40.0                 Fall   
51              100.0               Spring   
59                NaN               Summer   
87                NaN               Winter   
193              47.0               Spring   
...               ...                  ...   
3876              NaN               Winter   
3903             93.0               Summer   
3922             54.0               Spring

In [40]:
# df[((np.abs(stats.zscore(df[numericals], nan_policy='omit')) < 3) | (np.abs(stats.zscore(df[numericals], nan_policy='omit')).isnull())).all(axis=1)]

If we track outliers based on a $|zscore| >= 3$ metric, then that takes out 317 rows of the data!

In [41]:
317 / df.shape[0]

0.08005050505050505

8% outliers based on zscore

But, the data is mostly skewed, so perhaps we can try something other than zscore for outlier detection.  
How about based on IQR?

In [42]:
Q1 = df[numericals].quantile(0.25)
Q3 = df[numericals].quantile(0.75)
IQR = Q3 - Q1

In [43]:
df[((df[numericals] < (Q1 - 1.5 * IQR)) | (df[numericals] > (Q3 + 1.5 * IQR))).any(axis=1)]

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
5     001f3379                    Spring               13                1   
6     0038ba98                      Fall               10                0   
21    00f574e9                    Summer               16                0   
22    01085eb3                      Fall               12                0   
32    01d24f60                      Fall               16                0   
...        ...                       ...              ...              ...   
3935  fdd9799f                    Spring               15                0   
3938  fe102a25                    Summer                5                0   
3945  fe9c71d8                    Winter                9                0   
3946  feb53635                    Winter               14                0   
3952  ff60112d                    Summer               15                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
5         Winter             50.0          Summer     22.279952   
6            NaN              NaN            Fall     19.660760   
21           NaN              NaN          Summer     29.285856   
22        Winter             58.0            Fall     34.187282   
32           NaN              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3935      Summer             48.0          Summer     29.404637   
3938      Winter             61.0          Summer     18.278557   
3945      Spring             80.0          Winter     15.891137   
3946      Summer             70.0          Spring     19.036779   
3952      Spring             40.0          Winter     26.364710   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
5               59.50            112.2  ...             1.0             2.0   
6               55.00             84.6  ...             4.0             1.0   
21              71.00            210.0  ...             2.0             1.0   
22              60.50            178.0  ...             1.0             1.0   
32                NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3935            66.00            182.2  ...             2.0             4.0   
3938            43.50             49.2  ...             3.0             0.0   
3945            56.75             72.8  ...             1.0             1.0   
3946            69.50            130.8  ...             4.0             4.0   
3952            70.50            186.4  ...             1.0             1.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
5                1.0               34.0     Summer              40.0   
6                0.0               20.0     Winter              27.0   
21               0.0               30.0        NaN               NaN   
22               0.0               30.0       Fall              53.0   
32               NaN                NaN     Winter              67.0   
...              ...                ...        ...               ...   
3935             2.0               45.0     Summer              40.0   
3938             0.0               34.0        NaN               NaN   
3945             1.0               29.0     Winter              33.0   
3946             4.0               77.0     Spring              51.0   
3952             1.0               35.0     Spring              66.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
5                56.0               Spring   
6                40.0                 Fall   
21                NaN               Summer   
22               73.0                 Fall   
32               91.0                 Fall   
...               ...                  ...   
3935             56.0               Spring   
3938              NaN               Summer   
3945             47.0               Winter

In [44]:
784 / df.shape[0]

0.19797979797979798

20% outliers according to using the IQR!  
That's a lot of the data considered as outliers, perhaps we should widen the outlier fence boundaries

In [45]:
df[((df[numericals] < (Q1 - 2.5 * IQR)) | (df[numericals] > (Q3 + 2.5 * IQR))).any(axis=1)]

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
6     0038ba98                      Fall               10                0   
22    01085eb3                      Fall               12                0   
51    02cebf33                    Spring               12                1   
59    0380e709                    Summer               17                0   
144   0844aee7                    Summer                8                0   
...        ...                       ...              ...              ...   
3844  f7fc97cf                    Spring               14                0   
3863  f8ff0bc8                      Fall               18                0   
3922  fcd9a6bb                    Spring                6                0   
3946  feb53635                    Winter               14                0   
3952  ff60112d                    Summer               15                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
6            NaN              NaN            Fall     19.660760   
22        Winter             58.0            Fall     34.187282   
51        Spring             77.0          Summer     19.751301   
59           NaN              NaN          Summer     24.446886   
144         Fall             61.0            Fall     14.706858   
...          ...              ...             ...           ...   
3844      Spring             75.0          Spring     23.752311   
3863      Spring             62.0          Winter     26.429967   
3922      Summer             69.0          Spring     16.195145   
3946      Summer             70.0          Spring     19.036779   
3952      Spring             40.0          Winter     26.364710   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
6                55.0             84.6  ...             4.0             1.0   
22               60.5            178.0  ...             1.0             1.0   
51               62.0            108.0  ...             3.0             3.0   
59               68.0            160.8  ...             4.0             2.0   
144              48.0             48.2  ...             1.0             1.0   
...               ...              ...  ...             ...             ...   
3844             69.5            163.2  ...             1.0             1.0   
3863             71.5            192.2  ...             1.0             1.0   
3922             44.0             44.6  ...             1.0             0.0   
3946             69.5            130.8  ...             4.0             4.0   
3952             70.5            186.4  ...             1.0             1.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
6                0.0               20.0     Winter              27.0   
22               0.0               30.0       Fall              53.0   
51               3.0               46.0     Summer              82.0   
59               2.0               56.0        NaN               NaN   
144              1.0               32.0       Fall              32.0   
...              ...                ...        ...               ...   
3844             2.0               23.0     Spring              35.0   
3863             2.0               24.0     Winter              42.0   
3922             0.0               14.0     Spring              38.0   
3946             4.0               77.0     Spring              51.0   
3952             1.0               35.0     Spring              66.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
6                40.0                 Fall   
22               73.0                 Fall   
51              100.0               Spring   
59                NaN               Summer   
144              46.0               Summer   
...               ...                  ...   
3844             50.0               Spring   
3863             59.0                 Fall   
3922             54.0               Spring

In [46]:
274 / df.shape[0]

0.0691919191919192

Let's remove this 7% of the data that is extreme

In [47]:
df = df[~(((df[numericals] < (Q1 - 2.5 * IQR)) | (df[numericals] > (Q3 + 2.5 * IQR))).any(axis=1))]
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1            NaN              NaN            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956         NaN              NaN          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959         NaN              NaN          Winter           NaN   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0                46.0             50.8  ...             4.0             2.0   
1                48.0             46.0  ...             0.0             0.0   
2                56.5             75.6  ...             2.0             1.0   
3                56.0             81.6  ...             3.0             4.0   
4                 NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3955             59.5             82.4  ...             1.0             1.0   
3956             53.5             76.4  ...             NaN             NaN   
3957             60.0            109.8  ...             1.0             0.0   
3958             70.7             87.0  ...             1.0             1.0   
3959              NaN              NaN  ...             NaN             NaN   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
0                4.0               55.0        NaN               NaN   
1                0.0                0.0       Fall              46.0   
2                1.0               28.0       Fall              38.0   
3                1.0               44.0     Summer              31.0   
4                NaN                NaN        NaN               NaN   
...              ...                ...        ...               ...   
3955             0.0               32.0     Winter              35.0   
3956             NaN                NaN        NaN               NaN   
3957             1.0               31.0     Winter              56.0   
3958             1.0               19.0     Spring              33.0   
3959             NaN                NaN        NaN               NaN   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                 NaN                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                 NaN                  NaN   
...               ...                  ...   
3955             50.0                 Fall   
3956              NaN               Winter   
3957             77.0                 Fall

## Impute NaN values with random values from the same column

In [48]:
df.isnull().sum().sum()

111789

Let's impute with random values from the same column

In [49]:
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1            NaN              NaN            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer              NaN             NaN           NaN   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956         NaN              NaN          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959         NaN              NaN          Winter           NaN   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0                46.0             50.8  ...             4.0             2.0   
1                48.0             46.0  ...             0.0             0.0   
2                56.5             75.6  ...             2.0             1.0   
3                56.0             81.6  ...             3.0             4.0   
4                 NaN              NaN  ...             NaN             NaN   
...               ...              ...  ...             ...             ...   
3955             59.5             82.4  ...             1.0             1.0   
3956             53.5             76.4  ...             NaN             NaN   
3957             60.0            109.8  ...             1.0             0.0   
3958             70.7             87.0  ...             1.0             1.0   
3959              NaN              NaN  ...             NaN             NaN   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
0                4.0               55.0        NaN               NaN   
1                0.0                0.0       Fall              46.0   
2                1.0               28.0       Fall              38.0   
3                1.0               44.0     Summer              31.0   
4                NaN                NaN        NaN               NaN   
...              ...                ...        ...               ...   
3955             0.0               32.0     Winter              35.0   
3956             NaN                NaN        NaN               NaN   
3957             1.0               31.0     Winter              56.0   
3958             1.0               19.0     Spring              33.0   
3959             NaN                NaN        NaN               NaN   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                 NaN                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                 NaN                  NaN   
...               ...                  ...   
3955             50.0                 Fall   
3956              NaN               Winter   
3957             77.0                 Fall

In [50]:
for col in df.columns:
    choices = df[col].dropna().values
    
    nan_indices = df[col].isnull()
    df.loc[nan_indices, col] = np.random.choice(choices, size=nan_indices.sum())

In [51]:
df.isnull().sum().sum()

0

In [52]:
df

id Basic_Demos-Enroll_Season  Basic_Demos-Age  Basic_Demos-Sex  \
0     00008ff9                      Fall                5                0   
1     000fd460                    Summer                9                0   
2     00105258                    Summer               10                1   
3     00115b9f                    Winter                9                0   
4     0016bb22                    Spring               18                1   
...        ...                       ...              ...              ...   
3955  ff8a2de4                      Fall               13                0   
3956  ffa9794a                    Winter               10                0   
3957  ffcd4dbd                      Fall               11                0   
3958  ffed1dd5                    Spring               13                0   
3959  ffef538e                    Spring               11                0   

     CGAS-Season  CGAS-CGAS_Score Physical-Season  Physical-BMI  \
0         Winter             51.0            Fall     16.877316   
1           Fall             78.0            Fall     14.035590   
2           Fall             71.0            Fall     16.648696   
3           Fall             71.0          Summer     18.292347   
4         Summer             95.0          Spring     13.894456   
...          ...              ...             ...           ...   
3955      Spring             60.0            Fall     16.362460   
3956      Summer             72.0          Spring     18.764678   
3957      Spring             68.0          Winter     21.441500   
3958      Spring             70.0          Winter     12.235895   
3959      Winter             55.0          Winter     23.128212   

      Physical-Height  Physical-Weight  ...  PCIAT-PCIAT_18  PCIAT-PCIAT_19  \
0               46.00             50.8  ...             4.0             2.0   
1               48.00             46.0  ...             0.0             0.0   
2               56.50             75.6  ...             2.0             1.0   
3               56.00             81.6  ...             3.0             4.0   
4               48.50             62.6  ...             0.0             1.0   
...               ...              ...  ...             ...             ...   
3955            59.50             82.4  ...             1.0             1.0   
3956            53.50             76.4  ...             0.0             1.0   
3957            60.00            109.8  ...             1.0             0.0   
3958            70.70             87.0  ...             1.0             1.0   
3959            62.75             63.4  ...             2.0             1.0   

      PCIAT-PCIAT_20  PCIAT-PCIAT_Total SDS-Season SDS-SDS_Total_Raw  \
0                4.0               55.0     Summer              32.0   
1                0.0                0.0       Fall              46.0   
2                1.0               28.0       Fall              38.0   
3                1.0               44.0     Summer              31.0   
4                1.0               30.0       Fall              35.0   
...              ...                ...        ...               ...   
3955             0.0               32.0     Winter              35.0   
3956             0.0               19.0     Summer              38.0   
3957             1.0               31.0     Winter              56.0   
3958             1.0               19.0     Spring              33.0   
3959             1.0                0.0     Spring              71.0   

      SDS-SDS_Total_T  PreInt_EduHx-Season  \
0                59.0                 Fall   
1                64.0               Summer   
2                54.0               Summer   
3                45.0               Winter   
4                47.0                 Fall   
...               ...                  ...   
3955             50.0                 Fall   
3956             53.0               Winter   
3957             77.0                 Fall

# One Hot encoding

In [53]:
categoricals

Index(['id', 'Basic_Demos-Enroll_Season', 'CGAS-Season', 'Physical-Season',
       'Fitness_Endurance-Season', 'FGC-Season', 'BIA-Season', 'PAQ_C-Season',
       'PCIAT-Season', 'SDS-Season', 'PreInt_EduHx-Season'],
      dtype='object', name='Field')

Excluding the id, everything else is season data, so let's one hot encode them

In [54]:
categoricals = categoricals.drop('id')

In [55]:
from sklearn.preprocessing import OneHotEncoder

In [56]:
# Initialize OneHotEncoder
encoder = OneHotEncoder(sparse_output=False)
one_hot_encoded = encoder.fit_transform(df[categoricals])

one_hot_df = pd.DataFrame(one_hot_encoded,
                          columns=encoder.get_feature_names_out(categoricals))

print(one_hot_df.shape)
print(df.shape)

df.reset_index(drop=True, inplace=True)
one_hot_df.reset_index(drop=True, inplace=True)
df = pd.concat([df.drop(categoricals, axis=1), one_hot_df], axis=1)

print(df.shape)

df

(3686, 40)
(3686, 77)
(3686, 107)


id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
0     00008ff9                5                0             51.0   
1     000fd460                9                0             78.0   
2     00105258               10                1             71.0   
3     00115b9f                9                0             71.0   
4     0016bb22               18                1             95.0   
...        ...              ...              ...              ...   
3681  ff8a2de4               13                0             60.0   
3682  ffa9794a               10                0             72.0   
3683  ffcd4dbd               11                0             68.0   
3684  ffed1dd5               13                0             70.0   
3685  ffef538e               11                0             55.0   

      Physical-BMI  Physical-Height  Physical-Weight  \
0        16.877316            46.00             50.8   
1        14.035590            48.00             46.0   
2        16.648696            56.50             75.6   
3        18.292347            56.00             81.6   
4        13.894456            48.50             62.6   
...            ...              ...              ...   
3681     16.362460            59.50             82.4   
3682     18.764678            53.50             76.4   
3683     21.441500            60.00            109.8   
3684     12.235895            70.70             87.0   
3685     23.128212            62.75             63.4   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
0                             36.0                   76.0                69.0   
1                             22.0                   75.0                70.0   
2                             23.0                   65.0                94.0   
3                             26.0                   60.0                97.0   
4                             21.0                   66.0                83.0   
...                            ...                    ...                 ...   
3681                          24.0                   71.0                70.0   
3682                          27.0                   60.0                78.0   
3683                          34.0                   79.0                99.0   
3684                          35.0                   59.0                61.0   
3685                          28.0                   60.0               106.0   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
0     ...                  0.0                  0.0              0.0   
1     ...                  0.0                  0.0              1.0   
2     ...                  0.0                  0.0              1.0   
3     ...                  1.0                  0.0              0.0   
4     ...                  0.0                  1.0              1.0   
...   ...                  ...                  ...              ...   
3681  ...                  0.0                  1.0              0.0   
3682  ...                  0.0                  0.0              0.0   
3683  ...                  0.0                  1.0              0.0   
3684  ...                  0.0                  0.0              0.0   
3685  ...                  0.0                  0.0              0.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
0                   0.0                1.0                0.0   
1                   0.0                0.0                0.0   
2                   0.0                0.0                0.0   
3                   0.0                1.0                0.0   
4                   0.0                0.0                0.0   
...                 ...                ...                ...   
3681                0.0                0.0                1.0   
3682                0.0                1.0                0.0   
3683                0.0                0.0                1.0   
3684                1.

Let's check it

In [57]:
df.iloc[:,-4:]

PreInt_EduHx-Season_Fall  PreInt_EduHx-Season_Spring  \
0                          1.0                         0.0   
1                          0.0                         0.0   
2                          0.0                         0.0   
3                          0.0                         0.0   
4                          1.0                         0.0   
...                        ...                         ...   
3681                       1.0                         0.0   
3682                       0.0                         0.0   
3683                       1.0                         0.0   
3684                       0.0                         1.0   
3685                       0.0                         1.0   

      PreInt_EduHx-Season_Summer  PreInt_EduHx-Season_Winter  
0                            0.0                         0.0  
1                            1.0                         0.0  
2                            1.0                         0.0  
3                            0.0                         1.0  
4                            0.0                         0.0  
...                          ...                         ...  
3681                         0.0                         0.0  
3682                         0.0                         1.0  
3683                         0.0                         0.0  
3684                         0.0                         0.0  
3685                         0.0                         0.0  

[3686 rows x 4 columns]

# Normalization

We'll use min max normalization since the data is skewed.

We will first split the data into train and test  
Then we'll scale the numerical columns afterward

In [58]:
from sklearn.preprocessing import MinMaxScaler

In [60]:
# shuffle the data!
df = df.sample(frac=1)
df

id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
2857  c47452e3                8                0             80.0   
2280  99f3be05                6                0             62.0   
2114  8ef72e3e                6                1             52.0   
1109  48d285f4                5                1             68.0   
2342  9e7860ef               11                1             81.0   
...        ...              ...              ...              ...   
1338  5896b610                9                1             80.0   
1017  41d48900               12                1             61.0   
3450  eefeb431               12                1             75.0   
0     00008ff9                5                0             51.0   
1435  5f284ef1               22                1             55.0   

      Physical-BMI  Physical-Height  Physical-Weight  \
2857     17.791975            54.00             73.8   
2280     14.346939            42.00             36.0   
2114     14.789189            46.25             45.0   
1109     25.004665            49.00             85.4   
2342     20.251029            54.00             84.0   
...            ...              ...              ...   
1338     23.775365            53.00             95.0   
1017     21.519611            60.00            110.2   
3450     25.542333            52.00             55.8   
0        16.877316            46.00             50.8   
1435     24.835584            62.50            138.0   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
2857                          24.0                   71.0                64.0   
2280                          26.0                   73.0                87.0   
2114                          21.0                   66.0                96.0   
1109                          24.0                   66.0                84.0   
2342                          20.0                   67.0                99.0   
...                            ...                    ...                 ...   
1338                          28.0                   86.0                68.0   
1017                          25.0                   61.0                96.0   
3450                          37.0                   73.0                75.0   
0                             36.0                   76.0                69.0   
1435                          35.0                   76.0                71.0   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
2857  ...                  1.0                  0.0              0.0   
2280  ...                  1.0                  0.0              0.0   
2114  ...                  1.0                  0.0              0.0   
1109  ...                  0.0                  0.0              0.0   
2342  ...                  0.0                  1.0              1.0   
...   ...                  ...                  ...              ...   
1338  ...                  1.0                  0.0              1.0   
1017  ...                  1.0                  0.0              0.0   
3450  ...                  0.0                  0.0              0.0   
0     ...                  0.0                  0.0              0.0   
1435  ...                  1.0                  0.0              0.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
2857                0.0                0.0                1.0   
2280                0.0                1.0                0.0   
2114                1.0                0.0                0.0   
1109                0.0                1.0                0.0   
2342                0.0                0.0                0.0   
...                 ...                ...                ...   
1338                0.0                0.0                0.0   
1017                0.0                1.0                0.0   
3450                1.0                0.0                0.0   
0                   0.

In [109]:
dfTest = df.iloc[:df.shape[0] // 10]

In [110]:
dfTrain = df.iloc[df.shape[0] // 10:]

In [111]:
dfTrain[numericals] = dfTrain[numericals].astype('float64')
dfTest[numericals] = dfTest[numericals].astype('float64')

C:\Users\kerry\AppData\Local\Temp\ipykernel_10344\2330913363.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfTrain[numericals] = dfTrain[numericals].astype('float64')
C:\Users\kerry\AppData\Local\Temp\ipykernel_10344\2330913363.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfTest[numericals] = dfTest[numericals].astype('float64')


In [113]:
scaler = MinMaxScaler()
# df_normalized = pd.DataFrame(scaler.fit_transform(df[numericals]), columns=numericals)
dfTest.loc[:, numericals] = scaler.fit_transform(dfTest[numericals])

scaler = MinMaxScaler()
dfTrain.loc[:, numericals] = scaler.fit_transform(dfTrain[numericals])

In [114]:
dfTrain

id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
2314  9cb32129         0.235294                1         0.514286   
480   1d44d99b         0.058824                1         0.428571   
259   0f9b8e65         0.470588                0         0.542857   
1411  5e08237d         0.294118                0         0.714286   
21    01182ce3         0.470588                0         0.671429   
...        ...              ...              ...              ...   
1338  5896b610         0.235294                1         0.785714   
1017  41d48900         0.411765                1         0.514286   
3450  eefeb431         0.411765                1         0.714286   
0     00008ff9         0.000000                0         0.371429   
1435  5f284ef1         1.000000                1         0.428571   

      Physical-BMI  Physical-Height  Physical-Weight  \
2314      0.317714         0.390244         0.158041   
480       0.334239         0.243902         0.097967   
259       0.526801         0.548780         0.366913   
1411      0.494308         0.506098         0.317930   
21        0.444778         0.804878         0.498152   
...            ...              ...              ...   
1338      0.561110         0.414634         0.292052   
1017      0.478127         0.585366         0.362292   
3450      0.626111         0.390244         0.110906   
0         0.307351         0.243902         0.087800   
1435      0.600112         0.646341         0.490758   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
2314                      0.500000               0.388235            0.504950   
480                       0.208333               0.411765            0.485149   
259                       0.125000               0.529412            0.693069   
1411                      0.458333               0.447059            0.386139   
21                        0.208333               0.447059            0.425743   
...                            ...                    ...                 ...   
1338                      0.375000               0.682353            0.346535   
1017                      0.250000               0.388235            0.623762   
3450                      0.750000               0.529412            0.415842   
0                         0.708333               0.564706            0.356436   
1435                      0.666667               0.564706            0.376238   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
2314  ...                  0.0                  1.0              0.0   
480   ...                  0.0                  0.0              0.0   
259   ...                  0.0                  1.0              0.0   
1411  ...                  0.0                  0.0              0.0   
21    ...                  0.0                  0.0              0.0   
...   ...                  ...                  ...              ...   
1338  ...                  1.0                  0.0              1.0   
1017  ...                  1.0                  0.0              0.0   
3450  ...                  0.0                  0.0              0.0   
0     ...                  0.0                  0.0              0.0   
1435  ...                  1.0                  0.0              0.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
2314                0.0                0.0                1.0   
480                 0.0                0.0                1.0   
259                 0.0                0.0                1.0   
1411                1.0                0.0                0.0   
21                  0.0                0.0                1.0   
...                 ...                ...                ...   
1338                0.0                0.0                0.0   
1017                0.0                1.0                0.0   
3450                1.0                0.0                0.0   
0                   0.

In [115]:
dfTest

id  Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  \
2857  c47452e3         0.176471                0         0.765625   
2280  99f3be05         0.058824                0         0.484375   
2114  8ef72e3e         0.058824                1         0.328125   
1109  48d285f4         0.000000                1         0.578125   
2342  9e7860ef         0.352941                1         0.781250   
...        ...              ...              ...              ...   
2549  acde8e47         0.235294                1         0.687500   
1015  41bacefb         0.352941                1         0.484375   
1163  4bb9a8ff         0.176471                0         0.359375   
1531  6672e527         0.176471                1         0.578125   
2017  8823b020         0.411765                0         0.531250   

      Physical-BMI  Physical-Height  Physical-Weight  \
2857      0.332577         0.442953         0.251534   
2280      0.186303         0.120805         0.019632   
2114      0.205080         0.234899         0.074847   
1109      0.638824         0.308725         0.322699   
2342      0.436987         0.442953         0.314110   
...            ...              ...              ...   
2549      0.222470         0.469799         0.200000   
1015      0.404704         0.483221         0.322699   
1163      0.506981         0.397315         0.321472   
1531      0.274412         0.335570         0.157055   
2017      0.367739         0.577181         0.364417   

      Physical-Waist_Circumference  Physical-Diastolic_BP  Physical-HeartRate  \
2857                      0.208333               0.408451            0.364706   
2280                      0.291667               0.436620            0.635294   
2114                      0.083333               0.338028            0.741176   
1109                      0.208333               0.338028            0.600000   
2342                      0.041667               0.352113            0.776471   
...                            ...                    ...                 ...   
2549                      0.208333               0.436620            0.752941   
1015                      0.416667               0.295775            0.658824   
1163                      0.337500               0.647887            0.694118   
1531                      0.166667               0.464789            0.611765   
2017                      0.333333               0.380282            0.623529   

      ...  PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
2857  ...                  1.0                  0.0              0.0   
2280  ...                  1.0                  0.0              0.0   
2114  ...                  1.0                  0.0              0.0   
1109  ...                  0.0                  0.0              0.0   
2342  ...                  0.0                  1.0              1.0   
...   ...                  ...                  ...              ...   
2549  ...                  0.0                  1.0              0.0   
1015  ...                  0.0                  0.0              0.0   
1163  ...                  1.0                  0.0              0.0   
1531  ...                  0.0                  1.0              0.0   
2017  ...                  0.0                  1.0              0.0   

      SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
2857                0.0                0.0                1.0   
2280                0.0                1.0                0.0   
2114                1.0                0.0                0.0   
1109                0.0                1.0                0.0   
2342                0.0                0.0                0.0   
...                 ...                ...                ...   
2549                0.0                0.0                1.0   
1015                1.0                0.0                0.0   
1163                1.0                0.0                0.0   
1531                0.

In [116]:
dfTrain.describe()

Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  Physical-BMI  \
count      3318.000000      3318.000000      3318.000000   3318.000000   
mean          0.308549         0.374623         0.574774      0.378531   
std           0.203515         0.484098         0.169680      0.148137   
min           0.000000         0.000000         0.000000      0.000000   
25%           0.176471         0.000000         0.500000      0.268413   
50%           0.294118         0.000000         0.571429      0.337942   
75%           0.411765         1.000000         0.714286      0.455300   
max           1.000000         1.000000         1.000000      1.000000   

       Physical-Height  Physical-Weight  Physical-Waist_Circumference  \
count      3318.000000      3318.000000                   3318.000000   
mean          0.474658         0.249102                      0.322380   
std           0.173712         0.166143                      0.196651   
min           0.000000         0.000000                      0.000000   
25%           0.341463         0.118530                      0.166667   
50%           0.452439         0.202403                      0.291667   
75%           0.609756         0.353974                      0.458333   
max           1.000000         1.000000                      1.000000   

       Physical-Diastolic_BP  Physical-HeartRate  Physical-Systolic_BP  ...  \
count            3318.000000         3318.000000           3318.000000  ...   
mean                0.479619            0.483597              0.500423  ...   
std                 0.136661            0.133552              0.135693  ...   
min                 0.000000            0.000000              0.000000  ...   
25%                 0.388235            0.396040              0.413551  ...   
50%                 0.458824            0.475248              0.485981  ...   
75%                 0.552941            0.574257              0.570093  ...   
max                 1.000000            1.000000              1.000000  ...   

       PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
count          3318.000000          3318.000000      3318.000000   
mean              0.241410             0.229054         0.242013   
std               0.428003             0.420287         0.428367   
min               0.000000             0.000000         0.000000   
25%               0.000000             0.000000         0.000000   
50%               0.000000             0.000000         0.000000   
75%               0.000000             0.000000         0.000000   
max               1.000000             1.000000         1.000000   

       SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
count        3318.000000        3318.000000        3318.000000   
mean            0.276371           0.236287           0.245329   
std             0.447270           0.424864           0.430347   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           0.000000           0.000000   
max             1.000000           1.000000           1.000000   

       PreInt_EduHx-Season_Fall  PreInt_EduHx-Season_Spring  \
count               3318.000000                 3318.000000   
mean                   0.226341                    0.281495   
std                    0.418526                    0.449796   
min                    0.000000                    0.000000   
25%                    0.000000                    0.000000   
50%                    0.000000                    0.000000   
75%                    0.000000                    1.000000   
max                    1.000000                    1.000000   

       PreInt_EduHx-Season_Summer  PreInt_EduHx-Season_Winter  
count                 3318.000000                 3318.000000  
mean                     0.232068                    0.260096  
st

In [117]:
dfTest.describe()

Basic_Demos-Age  Basic_Demos-Sex  CGAS-CGAS_Score  Physical-BMI  \
count       368.000000       368.000000       368.000000    368.000000   
mean          0.316017         0.391304         0.546281      0.379172   
std           0.204002         0.488707         0.191675      0.170969   
min           0.000000         0.000000         0.000000      0.000000   
25%           0.176471         0.000000         0.453125      0.255031   
50%           0.294118         0.000000         0.531250      0.334604   
75%           0.411765         1.000000         0.687500      0.468663   
max           1.000000         1.000000         1.000000      1.000000   

       Physical-Height  Physical-Weight  Physical-Waist_Circumference  \
count       368.000000       368.000000                    368.000000   
mean          0.480603         0.323054                      0.331273   
std           0.188076         0.204046                      0.200803   
min           0.000000         0.000000                      0.000000   
25%           0.335570         0.164417                      0.166667   
50%           0.456376         0.264417                      0.291667   
75%           0.630872         0.465951                      0.458333   
max           1.000000         1.000000                      1.000000   

       Physical-Diastolic_BP  Physical-HeartRate  Physical-Systolic_BP  ...  \
count             368.000000          368.000000            368.000000  ...   
mean                0.370139            0.562596              0.507655  ...   
std                 0.153829            0.153159              0.148447  ...   
min                 0.000000            0.000000              0.000000  ...   
25%                 0.267606            0.458824              0.419355  ...   
50%                 0.352113            0.564706              0.483871  ...   
75%                 0.450704            0.670588              0.580645  ...   
max                 1.000000            1.000000              1.000000  ...   

       PCIAT-Season_Summer  PCIAT-Season_Winter  SDS-Season_Fall  \
count           368.000000           368.000000       368.000000   
mean              0.244565             0.230978         0.217391   
std               0.430414             0.422032         0.413033   
min               0.000000             0.000000         0.000000   
25%               0.000000             0.000000         0.000000   
50%               0.000000             0.000000         0.000000   
75%               0.000000             0.000000         0.000000   
max               1.000000             1.000000         1.000000   

       SDS-Season_Spring  SDS-Season_Summer  SDS-Season_Winter  \
count         368.000000         368.000000         368.000000   
mean            0.293478           0.247283           0.241848   
std             0.455976           0.432020           0.428786   
min             0.000000           0.000000           0.000000   
25%             0.000000           0.000000           0.000000   
50%             0.000000           0.000000           0.000000   
75%             1.000000           0.000000           0.000000   
max             1.000000           1.000000           1.000000   

       PreInt_EduHx-Season_Fall  PreInt_EduHx-Season_Spring  \
count                368.000000                  368.000000   
mean                   0.263587                    0.252717   
std                    0.441178                    0.435162   
min                    0.000000                    0.000000   
25%                    0.000000                    0.000000   
50%                    0.000000                    0.000000   
75%                    1.000000                    1.000000   
max                    1.000000                    1.000000   

       PreInt_EduHx-Season_Summer  PreInt_EduHx-Season_Winter  
count                  368.000000                  368.000000  
mean                     0.217391                    0.266304  
st

Export to csvs!

In [118]:
dfTrain.to_csv('content/dataTrain.csv', index=False)
dfTest.to_csv('content/dataTest.csv', index=False)